### PR TITLE
[Core] Add engine option to return only deltas or final output

### PR DESCRIPTION
**Pull request by njhill**
The `LLMEngine` and `AsyncLLMEngine` APIs will currently return/stream cumulative outputs and prompt-related data for all sequences at every step.

This is more data than needed for `LLM.generate` or the OpenAI server APIs:
- For LLM.generate and non-streaming APIs we only need the final output
- For streaming APIs we only require deltas

This PR adds an `output_kind` parameter to `SamplingParams` with an enum value of either `CUMULATIVE`, `DELTA`, or `FINAL_ONLY`.

In the `DELTA` case, data associated with the prompt (prompt token ids, logits, etc.) is returned only in the first output message(s).

It will reduce the number of objects that need to be constructed at each step, and the amount of data to be serialized to return to the newly-decoupled front-end API process.

### CODE REVIEW

**Comment by njhill** [comment]
This is ready, I'm just planning to add a couple of new tests for it so will hold off adding the ready label for that.

File: vllm/outputs.py [hunk 1]
```diff
@@ -113,19 +114,30 @@ def __init__(
         self.encoder_prompt_token_ids = encoder_prompt_token_ids
 
     @classmethod
-    def from_seq_group(cls, seq_group: SequenceGroup) -> "RequestOutput":
-        if seq_group.sampling_params is None:
+    def from_seq_group(
+        cls,
+        seq_group: SequenceGroup,
+        prior_output_lens: Dict[int, Tuple[int, int]],
+        include_prompt: bool = True,
```
Comment (joerunde) [inline vllm/outputs.py:121]: Can this parameter be inferred by checking the output kind and prior output lens? We only don't want the prompt if it's a delta response and the prior output lens are non-zero, right?

Comment (njhill) [reply vllm/outputs.py:121]: @joerunde that's effectively what's already done but yes good idea, can encapsulate it within the method

Comment (njhill) [reply vllm/outputs.py]: Thanks @joerunde that was a nice simplification

**Comment by njhill** [comment]
Blocked on some testing issues that I'm trying to solve in https://github.com/vllm-project/vllm/pull/7565

File: vllm/entrypoints/openai/serving_chat.py [hunk 2]
```diff
@@ -264,15 +263,36 @@ async def chat_completion_stream_generator(
...
+
         tool_parser: Optional[ToolParser] = self.tool_parser(
             tokenizer) if self.tool_parser else None
 
+        if isinstance(request.tool_choice, ChatCompletionNamedToolChoiceParam):
+            tool_choice_function_name = request.tool_choice.function.name
+        else:
+            tool_choice_function_name = None
+
+        tool_choice_auto = (
```
Comment (alexm-neuralmagic) [inline vllm/entrypoints/openai/serving_chat.py:279]: A quick documentation would be helpful here.

File: vllm/outputs.py [hunk 3]
```diff
@@ -113,19 +114,26 @@ def __init__(
...
-    def from_seq_group(cls, seq_group: SequenceGroup) -> "RequestOutput":
-        if seq_group.sampling_params is None:
+    def from_seq_group(cls,
+                       seq_group: SequenceGroup) -> Optional["RequestOutput"]:
+        sampling_params = seq_group.sampling_params
+        if sampling_params is None:
             raise ValueError(
                 "Sampling parameters are missing for a CompletionRequest.")
+        finished = seq_group.is_finished()
+        if sampling_params.output_kind == RequestOutputKind.FINAL_ONLY and (
```
Comment (alexm-neuralmagic) [inline vllm/outputs.py:124]: Good idea to use sampling_params as the place for output_kind, makes everything simpler.

File: vllm/outputs.py [hunk 4]
```diff
@@ -135,26 +143,50 @@ def from_seq_group(cls, seq_group: SequenceGroup) -> "RequestOutput":
...
+        outputs = []
+        include_prompt = True
+        for seq in top_n_seqs:
+            output_text = seq.get_output_text_to_return(text_buffer_length)
+            output_logprobs = seq.output_logprobs if include_logprobs else None
+
+            output_token_ids: GenericSequence[int]
+            if deltas:
+                output_token_ids = seq.data.last_appended_tokens
+                seq.data.last_appended_tokens = []
```
Comment (alexm-neuralmagic) [inline vllm/outputs.py:159]: I would use .clear() here since it saves on deallocation=>allocation.

File: vllm/outputs.py [hunk 5]
```diff
@@ -135,26 +143,50 @@ def from_seq_group(cls, seq_group: SequenceGroup) -> "RequestOutput":
...
-                SequenceStatus.get_finished_reason(seq.status),
-                seq.stop_reason) for seq in top_n_seqs
-        ]
+        include_logprobs = sampling_params.logprobs is not None
+        text_buffer_length = sampling_params.output_text_buffer_length
+        deltas = sampling_params.output_kind == RequestOutputKind.DELTA
+
+        outputs = []
+        include_prompt = True
+        for seq in top_n_seqs:
```
Comment (alexm-neuralmagic) [inline vllm/outputs.py:152]: Good that it also includes the beam search case

File: vllm/sequence.py [hunk 6]
```diff
@@ -451,12 +457,17 @@ def prompt_adapter_id(self) -> int:
...
                         if self.prompt_adapter_request else 0
 
-    def get_output_text_to_return(self, buffer_length: int):
+    def get_output_text_to_return(self, buffer_length: int) -> str:
         # We return the full output text if the sequence is finished.
         truncate = buffer_length and not self.is_finished()
         return self.output_text[:-buffer_length] if truncate else (
             self.output_text)
 
+    def get_output_text_to_return_len(self, buffer_length: int) -> int:
```
Comment (alexm-neuralmagic) [inline vllm/sequence.py:466]: Is this function used?

**Review by alexm-redhat — APPROVED** [review]
@njhill thanks for adding the delta request output support, this is really helpful! Left some nit comments, but looks good overall

File: vllm/outputs.py (hunk 4 above)
Comment (njhill) [reply vllm/outputs.py:159]: But wouldn't this clear the list that we're setting in the `CompletionOutput` object that's getting returned?

File: vllm/sequence.py (hunk 6 above)
Comment (njhill) [reply vllm/sequence.py:466]: Ah, not any more, well spotted thanks!

File: vllm/outputs.py (hunk 4 above)
Comment (njhill) [reply vllm/outputs.py]: Resolving since this is no longer applicable after latest refactor.

**Comment by njhill** [comment]
Failed tests are unrelated flakes / already failing on main.

**Comment by alexm-neuralmagic** [comment]
Ok cool, LGTM

**Review by simon-mo — APPROVED** [review]
